readthedocs/proxito/views/hosting.py

```diff
@@ -104,44 +123,41 @@ def _resolve_resources(self):
                 project = unresolved_url.project
                 version = unresolved_url.version
                 filename = unresolved_url.filename
-                # This query should use a particular index:
-                # ``builds_build_version_id_state_date_success_12dfb214_idx``.
-                # Otherwise, if the index is not used, the query gets too slow.
-                build = (
-                    version.builds.filter(
-                        success=True,
-                        state=BUILD_STATE_FINISHED,
-                    )
-                    .select_related("project", "version")
-                    .first()
-                )
-
             except UnresolverError as exc:
                 # If an exception is raised and there is a ``project`` in the
                 # exception, it's a partial match. This could be because of an
                 # invalid URL path, but on a valid project domain. In this case, we
                 # continue with the ``project``, but without a ``version``.
                 # Otherwise, we return 404 NOT FOUND.
                 project = getattr(exc, "project", None)
-                if not project:
-                    raise Http404() from exc
-
         else:
-            project = Project.objects.filter(slug=project_slug).first()
-            version = (
-                Version.objects.filter(slug=version_slug, project=project)
-                .select_related("project")
-                .first()
+            # When not sending "url", we require "project-slug" and "version-slug".
+            project = get_object_or_404(Project, slug=project_slug)
+            version = get_object_or_404(project.versions.all(), slug=version_slug)
+
+        # A project is always required.
+        if not project:
+            return JsonResponse(
+                {"error": ClientError.PROJECT_NOT_FOUND},
+                status=404,
```

Comment:
This `return` doesn't work here. We need to either raise a not found exception from here, or move this `return` to the `get()` method.

Reply:
... or maybe it not required at all since the line 135 (`get_object_or_404`) will already raise the exception for it?

Reply:
Yeah, can't return here, will just raise a 404

> ... or maybe it not required at all since the line 135 (get_object_or_404) will already raise the exception for it?

when using a URL, the project can be none.